Release workload on a non-MNNVL node in a CD

jgehrcke opened the pull request:
Resolves https://github.com/NVIDIA/k8s-dra-driver-gpu/issues/625.

Slightly less trivial than I thought.

- Needs a special case `maxNodesPerIMEXDomain` for the group of non-MNNVL nodes (can become large)
- Always needs injection of CD details into CD daemon container (via CDI)

The group of nodes that does not perform MNNVL may be large. With this change, all of these nodes will contribute to the update-with-retry-on-conflict race, which is important to consider. I added a few code comments along the way.

I tested this manually, with this patch to simulate the scenario:

```
--- a/cmd/compute-domain-kubelet-plugin/nvlib.go
+++ b/cmd/compute-domain-kubelet-plugin/nvlib.go
@@ -197,6 +197,9 @@ func (l deviceLib) enumerateComputeDomainDaemons(config *Config) (AllocatableDev
 }
 
 func (l deviceLib) getCliqueID() (string, error) {
+
+       return "", nil
+
```

CD daemon log excerpt, from start-to-shutdown:
```
± kubectl logs -n nvidia-dra-driver-gpu   imex-channel-injection-p7m8n-2ghhj
I1007 11:24:04.713698       1 main.go:200] config: &{gb-nvl-043-compute06 49426eaf-63f5-4610-8928-306a99de8777 imex-channel-injection default  192.168.35.115 imex-channel-injection-p7m8n-2ghhj nvidia-dra-driver-gpu 18}
I1007 11:24:04.713783       1 main.go:209] no cliqueID: register with ComputeDomain, but do not run IMEX daemon
I1007 11:24:04.714026       1 dnsnames.go:199] Created static nodes config file with 18 DNS names using format compute-domain-daemon-%d
[...]
I1007 11:24:04.715152       1 process.go:175] Start watchdog
I1007 11:24:04.715185       1 main.go:338] wait for nodes update
I1007 11:24:04.715330       1 reflector.go:358] "Starting reflector" type="*v1beta1.ComputeDomain" resyncPeriod="10m0s" reflector="pkg/nvidia.com/informers/externalversions/factory.go:141"
I1007 11:24:04.715346       1 reflector.go:404] "Listing and watching" type="*v1beta1.ComputeDomain" reflector="pkg/nvidia.com/informers/externalversions/factory.go:141"
[...]
I1007 11:24:04.916398       1 computedomain.go:266] CD status does not contain node name 'gb-nvl-043-compute06' yet, try to insert myself: &{gb-nvl-043-compute06   0 NotReady}
[...]
I1007 11:24:04.919998       1 computedomain.go:288] Successfully updated CD
[...]
I1007 11:24:04.920268       1 main.go:350] empty cliqueID: do not start IMEX daemon
[...]
I1007 11:24:06.092166       1 podmanager.go:194] Successfully updated node gb-nvl-043-compute06 status to Ready
[...]
I1007 11:58:36.795592       1 main.go:89] Received SIGTERM, initiate shutdown
I1007 11:58:36.795671       1 main.go:341] shutdown: stop IMEXDaemonUpdateLoopWithDNSNames
I1007 11:58:36.795696       1 main.go:278] Terminated: IMEX daemon update task
I1007 11:58:36.795690       1 process.go:179] Watchdog: context canceled, attempt to stop child process
E1007 11:58:36.795727       1 main.go:288] watch failed, initiate shutdown: watchdog: stop failed: pm: stop failed: not started
I1007 11:58:36.795737       1 main.go:291] Terminated: process manager
[...]
I1007 11:58:36.803227       1 main.go:257] Terminated: controller task
I1007 11:58:36.803241       1 main.go:297] Exiting
````

With the `getCliqueID()` patch removed, I saw a passing test suite:

```text
15 tests, 0 failures in 118 seconds
```

jgehrcke commented on cmd/compute-domain-daemon/computedomain.go:
```diff
@@ -349,7 +371,7 @@ func (m *ComputeDomainManager) MaybePushNodesUpdate(cd *nvapi.ComputeDomain) {
 		m.updatedNodesChan <- cd.Status.Nodes
 	} else {
-		klog.Infof("IP set did not change")
+		klog.V(6).Infof("IP set did not change")
```
Seen along the way -- this message is logged at high frequency.

klueska commented on cmd/compute-domain-daemon/computedomain.go:
```diff
@@ -254,7 +259,11 @@ func (m *ComputeDomainManager) UpdateComputeDomainNodeInfo(ctx context.Context,
 			CliqueID: m.config.cliqueID,
 			Index:    nextIndex,
+			// Initialize as NotReady (will be updated by podmanager).
+			Status: nvapi.ComputeDomainStatusNotReady,
```
Sure. I would drop the comment though.

klueska commented on cmd/compute-domain-daemon/computedomain.go:
```diff
@@ -263,17 +272,22 @@ func (m *ComputeDomainManager) UpdateComputeDomainNodeInfo(ctx context.Context,
+	// Try to write update. May be in conflict with updates to the same list
+	// coming in from other nodes (in which case this update will be retried
+	// later). TODO: review conflict resolution methodology for the
+	// 10000-nodes-CD use case.
```
After each attempt there is an exponential backoff with some random jitter to stagger attempts from different nodes. We also (re)read the CD object just before attempting to write, in order to make the window as small as possible where a conflict might slip in. We would need to measure it, but I don't think the number of average conflicts would be nearly as large as you are suggesting.

klueska commented on cmd/compute-domain-kubelet-plugin/computedomain.go:
```diff
@@ -59,7 +59,8 @@ type ComputeDomainManager struct {
 type ComputeDomainDaemonSettings struct {
-	manager         *ComputeDomainManager
+	manager *ComputeDomainManager
+	// CD UUID
```
What is this comment meant to represent?

jgehrcke replied:
The this string carries the the CD UUID (as opposed to the CD name).

klueska commented on cmd/compute-domain-kubelet-plugin/device_state.go:
```diff
@@ -491,6 +492,29 @@ func (s *DeviceState) applyComputeDomainDaemonConfig(ctx context.Context, config
 	}
 
+	cd, err := s.computeDomainManager.GetComputeDomain(ctx, config.DomainID)
+	if err != nil {
+		return nil, fmt.Errorf("error getting compute domain: %w", err)
+	}
+	if cd == nil {
+		return nil, fmt.Errorf("compute domain not found: %s", config.DomainID)
+	}
+
+	// Always inject Compute Domain details into the CD daemon, in a
+	// heterogeneous CD clique ID below may be empty (while other CD details are
+	// set, and consumed by the CD daemon).
+	edits := &cdiapi.ContainerEdits{
+		ContainerEdits: &cdispec.ContainerEdits{
+			Env: []string{
+				fmt.Sprintf("CLIQUE_ID=%s", s.computeDomainManager.cliqueID),
+				fmt.Sprintf("COMPUTE_DOMAIN_UUID=%s", cd.UID),
+				fmt.Sprintf("COMPUTE_DOMAIN_NAME=%s", cd.Name),
+				fmt.Sprintf("COMPUTE_DOMAIN_NAMESPACE=%s", cd.Namespace),
+			},
+		},
+	}
```
Let's encapuslate this in a separate function on the `computeDomainManager` and rename computeDomainDaemonSettings.GetCDIContainerEdits to reflect that its specific to when IMEX is involved.

jgehrcke commented on cmd/compute-domain-daemon/computedomain.go:
```diff
@@ -221,6 +225,7 @@ func (m *ComputeDomainManager) UpdateComputeDomainNodeInfo(ctx context.Context,
 	var nodeInfo *nvapi.ComputeDomainNode
 
 	// Create a deep copy of the ComputeDomain to avoid modifying the original
+	// TODO: review for 10000-node-CD
```
No perceived issue.

Still, maybe we can do the copy later when we know that we have to perform a mutation.

jgehrcke commented on cmd/compute-domain-daemon/main.go:
```diff
@@ -302,6 +306,11 @@ func IMEXDaemonUpdateLoopWithIPs(ctx context.Context, controller *Controller, cl
 				return fmt.Errorf("writeNodesConfig failed: %w", err)
 			}
 
+			if cliqueID == "" {
+				klog.V(1).Infof("empty cliqueID: do not start IMEX daemon")
+				continue
+			}
+
```
Yeah. I agree (for real!).

I will look at this in a minute. If I see an elegant improvement, I will try it out. If not, I am in favor of leaving this as-is for now. Will report back.

jgehrcke replied:
I thought about this again, and didn't do any further change.

While this looks a bit ugly, it's conceptually safe and easy to reason about: we explicitly skip (only!) launching the child process. If we skip anything else we risk not doing something that in fact we want to do.

We of course _can_ try out other things, but the test cycle on this is a little slow -- I did patch `getCliqueID()` to return an empty string and then tested things out manually.

I am glad that things seem to work for now.

Let's streamline this in a follow-up, against an automatic test. I didn't get to building one yet; I think doing that is more important than various other things we may still talk about here. For building a test: maybe a special environment variable used by only the test suite may do it (to trigger this code path even if a clique ID is set on the system). Tracking building a test here: https://github.com/NVIDIA/k8s-dra-driver-gpu/issues/648.

jgehrcke commented on cmd/compute-domain-daemon/computedomain.go:
```diff
@@ -201,13 +203,15 @@ func (m *ComputeDomainManager) onAddOrUpdate(ctx context.Context, obj any) error
 	}
 
 	// Update node info in ComputeDomain
+	// Why are we doing this (only) upon receiving another update?
```
> When the informer first comes online it triggers an "add" on all of the existing objects that it receives from the API server.

Thank you! (I now remember that you've described this before)

jgehrcke commented:
@klueska thanks for review. I added responded to comments and added commits. Can you have another look? Before merge, I would squash commits.

jgehrcke commented:
I have rebased this branch on current head of main, resolved conficts, and also squashed commits.

To be sure the merge-to-main wouldn't be bad, I just tested this again manually with 
```
--- a/cmd/compute-domain-kubelet-plugin/nvlib.go
+++ b/cmd/compute-domain-kubelet-plugin/nvlib.go
@@ -197,6 +197,7 @@ func (l deviceLib) enumerateComputeDomainDaemons(config *Config) (AllocatableDev
 }
 
 func (l deviceLib) getCliqueID() (string, error) {
+       return "", nil
```

Works (CD daemon log):

```
...
15:32:24 ± kubectl logs -n nvidia-dra-driver-gpu   imex-channel-injection-rdx24-k9qg5
I1008 15:32:22.570420       1 main.go:200] config: &{gb-nvl-043-compute07 91d00090-848c-4953-8a25-ffe23d409e90 imex-channel-injection default  192.168.34.186 imex-channel-injection-rdx24-k9qg5 nvidia-dra-driver-gpu 18}
I1008 15:32:22.570486       1 main.go:209] no cliqueID: register with ComputeDomain, but do not run IMEX daemon
...
I1008 15:32:24.173219       1 podmanager.go:201] Successfully updated node gb-nvl-043-compute07 status to Ready
...
```

Workload got released:
```
$ kubectl get pods
NAME                     READY   STATUS    RESTARTS   AGE
imex-channel-injection   1/1     Running   0          3m20s
```

Is of course half-broken because we didn't perform CDI-based injection:
```
$ kubectl  logs imex-channel-injection 
ls: cannot access '/dev/nvidia-caps-imex-channels': No such file or directory
```

Let's land this now, (unless there's a rough problem in here). Let's do all non-functional changes in a follow-up. Specifically, let us sort out [finding a better encapsulation for the non-imex-daemon-mode](https://github.com/NVIDIA/k8s-dra-driver-gpu/pull/646#discussion_r2414226537) later.
